rfcs/20181016-optimizer-unification.md

```diff
@@ -0,0 +1,398 @@
+
+In addition, it is fairly confusing for users to have 2 distinct sets of optimizers with a different feature set.
+
+Thus we should seek to unify the Keras optimizer API and the TensorFlow optimizer API, by 1) extending the TensorFlow optimizer API, 2) replacing the tf.keras optimizers the upgraded TF optimizers.
+
+---
+
+## Objective
+
+- Unify `tf.train` and `tf.keras.optimizers` API:
+    - Make all TensorFlow optimizers JSON-serializable, and make it possible to save/restore their state.
+    - Make it possible to dynamically modify the value of the hyperparameters of all TensorFlow optimizers, in particular the learning rate.
+        - The current way to achieve dynamic learning rates is 1) use a LR tensor with built-in decay, 2) use a callable. Both of these approaches are limited (do not support fully-dynamic rates, e.g. adapting the rate based on the current loss decrease), and not intuitive. Doing `optimizer.lr = 0.2` at arbitrary points during training is eager-first and more user-friendly.
+- Have a single set of optimizers (same signatures, same objects, no wrappers), introduced as a new set of classes with an updated API, importable from `tf.keras.optimizers`. These optimizers would be based on the existing `tf.contrib.optimizer_v2` optimizers (which themselves are based on `tf.train optimizers`).
+
+
+The old optimizers will exist in tf.compat.v1 as-is.
+
+The known breaking changes are:
+- Due to name changes, old checkpoints would not be loadable with the new optimizers. This is opt-in: your checkpoint won't break until you start using the new optimizers in your code (you can always import the old optimizers from tf.compat.v1).
+- Some arguments are getting renamed.
+- The `use_locking` argument is removed.
+
+---
+
+## Design Proposal
+
+- Add a `get_config` method on every optimizer, as well as a `from_config` class method, to serialize / deserialize an optimizer (does not include weights value, i.e. state, but only includes hyperparameter values, i.e. the arguments that can be passed to the constructor).
+- Add a `get_weights` and `set_weights` method, to retrieve (or set) the optimizer’s state as a list of numpy arrays -- this is necessary for compatibility with the Keras API.
+- Add ability to set the values of optimizer hyperparameters (i.e. the arguments that can be passed to the constructor) at any point in the lifetime of the optimizer, without having to reinstantiate it. In particular this includes the ability to change the value of the learning rate.
+- Add support for gradient clipping by norm and by value.
+- Disable reusing a single optimizer instance across multiple graphs.
+- Move the optimizer classes to `tf.keras.optimizers`, with revise signatures (see details below).
+
+
+---
+
+## Detailed Design
```

Comment:
API design notes:
 * Could you clarify the behavior of get_weights and set_weights? It's unclear what is the effect when TF variables are involved. An example for an optimizer that tracks multiple shadow variables like Adam would be great. A link to an existing reference would work too.
 * A get_updates method would be extremely useful for tasks such as meta learning and imperative execution. It would be a pity to let this opportunity go by without adding one.